### PR TITLE
feat: refresh_xblock_skills management command

### PR DESCRIPTION
**Pull request by navinkarkera**
Adds command `refresh_xblock_skills` to update skills related to passed xblocks, or all xblocks in passed courses or all xblocks in all courses available from `get_course_metadata_provider`.

**Reviewers:**
- [x] @kaustavb12
- [x] @sameenfatima78 

**Test instructions:**

If you try out below commands in project directory, it should fail at ESME/lightcast API call due to incorrect URL and not before that. It should use test settings by default and use mock data for content. If not run `make test` once.

```bash
python manage.py migrate
python manage.py refresh_xblock_skills --xblock block-v1:edX+DemoX+Demo_Course+type@video+block@0b9e39477cf34507a7a48f74be381fdd
python manage.py refresh_xblock_skills --course some-id
python manage.py refresh_xblock_skills --all
```

To test `args-from-database` argument open django shell and run below commands:
```python
from taxonomy.models import RefreshXBlockSkillsConfig
config = RefreshXBlockSkillsConfig.get_solo()
config.arguments = ' --all --commit '
config.save()
```
Then run below command:
`python manage.py refresh_xblock_skills --args-from-database`

*Although, running above commands should cover the code in this MR, for end to end test with real metadata provider from course-discovery, I'll try to come up with test instructions once I figure out way to mock lightcast API or get free access to lightcast API*

**Merge checklist:**
- [x] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `make upgrade && make requirements` have been run to regenerate requirements
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
- [x] [Version](https://github.com/edx/taxonomy-connector/blob/master/taxonomy/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/taxonomy-connector/blob/master/CHANGELOG.rst) record added
- [ ] https://github.com/openedx/openedx-events/pull/143
- [ ] Update openedx-events dependency

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/taxonomy-connector/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/taxonomy-connector/actions), verify version has been pushed to [PyPI](https://pypi.org/project/taxonomy-connector/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [course-discovery](https://github.com/edx/course-discovery) to upgrade dependencies (including taxonomy-connector)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in course-discovery will look for the latest version in PyPi.

### CODE REVIEW

**Review by kaustavb12 — APPROVED** [review]
:+1:

Great job here @navinkarkera 
I have left a few minor comments, otherwise this looks good to go from my end.

- [x] I tested this: Tested the management command works as expected in devstack
- [x] I read through the code
- [ ] I checked for accessibility issues
- [x] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

File: taxonomy/management/commands/refresh_xblock_skills.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                )
+        else:
+            raise InvalidCommandOptionsError('Either course, xblock or --all argument must be provided.')
+
+        for course in courses:
+            course_key = course.get("key")
+            if not course_key:
+                LOGGER.error('[TAXONOMY] Invalid course metadata: missing attribute `key`')
+                continue
+            xblocks = xblock_provider.get_all_xblocks_in_course(course_key)
```
[the remark this answers is not in the document]
Comment (navinkarkera) [reply taxonomy/management/commands/refresh_xblock_skills.py:118]: Thanks for catching it. The only case xblocks will be empty for a valid course_key is that the course is empty, so raising error in that case does not make sense. Instead we can log an error and continue with the next course. The same should work if one of the course key is incorrect, we will log and process other courses.

File: taxonomy/management/commands/refresh_xblock_skills.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        )
+        parser.add_argument(
+            '--all',
+            action='store_true',
+            help=_('Create xblock skill mapping for all xblocks in all the courses.'),
+        )
+        parser.add_argument(
+            '--commit',
+            action='store_true',
+            default=False,
```
[the remark this answers is not in the document]
Comment (navinkarkera) [reply taxonomy/management/commands/refresh_xblock_skills.py:67]: Ohh, so we can replace these lines with just `action='store_false'`

File: taxonomy/management/commands/refresh_xblock_skills.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            help=u'Commits the skills to storage. '
```
[the remark this answers is not in the document]
Comment (navinkarkera) [reply taxonomy/management/commands/refresh_xblock_skills.py:68]: Missed it, thanks.

File: taxonomy/management/commands/refresh_xblock_skills.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            raise InvalidCommandOptionsError('Either course, xblock or --all argument must be provided.')
+
+        for course in courses:
+            course_key = course.get("key")
+            if not course_key:
+                LOGGER.error('[TAXONOMY] Invalid course metadata: missing attribute `key`')
+                continue
+            xblocks = xblock_provider.get_all_xblocks_in_course(course_key)
+            LOGGER.info('[TAXONOMY] Refresh xblocks skills process started for course: {course_key}.')
+            utils.refresh_product_skills(xblocks, options['commit'], self.product_type)
```
[the remark this answers is not in the document]
Comment (navinkarkera) [reply taxonomy/management/commands/refresh_xblock_skills.py:120]: `get_all_xblocks_in_course` returns an iterator which we process one by one in `refresh_product_skills` as the list of xblocks in a course can be huge, so we cannot append it to `xblocks_from_args`.

**Comment by navinkarkera** [comment]
@sameenfatima78 This is ready for your review.

File: taxonomy/management/commands/refresh_xblock_skills.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        LOGGER.info('[TAXONOMY] Refresh XBlock Skills. Options: [%s]', options)
+
+        courses = []
+        xblocks_from_args = []
+        xblock_provider = get_xblock_metadata_provider()
+        if options['all']:
+            courses = get_course_metadata_provider().get_all_courses()
+        elif options['course']:
+            courses = [{"key": course} for course in options['course']]
```
Comment (sameenfatima78) [inline taxonomy/management/commands/refresh_xblock_skills.py:103]: Are we supposed to pass a course key or a course id here? The usage examples have both. It should raise an exception if it's not in the required format. Would be helpful if we can add this info in the help text as well.

**Review by sameenfatima78 — COMMENTED** [review]
Just some minor feedback. Looks good otherwise.

File: taxonomy/management/commands/refresh_xblock_skills.py (hunk 5 above)
Comment (navinkarkera) [reply taxonomy/management/commands/refresh_xblock_skills.py:103]: @sameenfatima78 It requires a `course-key`.

>The usage examples have both.

Updated this:
`$ ./manage.py refresh_xblock_skills --course 'course-v1:edX+DemoX+Demo_Course' --course 'Course2_key' --commit`
to 
`$ ./manage.py refresh_xblock_skills --course 'course-v1:edX+DemoX+Demo_Course' --course 'course-v2:edX+DemoX+Demo_Course2' --commit`

>It should raise an exception if it's not in the required format.

Should we include `opaque_keys` as dependency and use `course_key = CourseKey.from_string(course_key_str)` as mentioned in 
https://github.com/openedx/edx-platform/wiki/Opaque-Keys-(Locators)#parse-course-ids

Comment (sameenfatima78) [reply taxonomy/management/commands/refresh_xblock_skills.py:103]: I think it's better to add a check right in the beginning before any service to service calls are made. We can add Opaque Keys as dependency here.

Comment (navinkarkera) [reply taxonomy/management/commands/refresh_xblock_skills.py:117]: @sameenfatima78 Done. Also validated xblock usage_keys.

**Review by sameenfatima78 — APPROVED** [review]

**Comment by navinkarkera** [comment]
@sameenfatima78 Rebased and ready to be merged.

**Comment by navinkarkera** [comment]
@sameenfatima78 Actually, we already merged another MR: #127 pointing to dev version of openedx-events and also the upstream MR is taking too long to be merged. So we can probably merge this as well and I'll create a new MR as soon as openedx-events is updated.

**Comment by openedx-webhooks** [comment]
@navinkarkera Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/taxonomy-connector&entry.867055334=https://github.com/openedx/taxonomy-connector/pull/129&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2022-12-07+13:10&entry.1917517419=2022-12-27+06:58&entry.2133058324=No) so we can improve your experience in the future.

**Comment by mphilbrick211** [comment]
Hi @navinkarkera! Is this ready to be reviewed/merged?

**Comment by navinkarkera** [comment]
@mphilbrick211 Yes, thank you.

**Comment by openedx-webhooks** [comment]
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/taxonomy-connector&entry.867055334=https://github.com/openedx/taxonomy-connector/pull/129&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2022-12-07+13:10&entry.1917517419=2022-12-29+08:37&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by sameenfatima78** [comment]
@navinkarkera Getting this error while publishing latest version to PyPI. https://github.com/openedx/taxonomy-connector/actions/runs/3799391856/jobs/6461895388
Screenshot 2022-12-29 at 1 51 30 PM

**Comment by navinkarkera** [comment]
@sameenfatima78 Ohh, it seems we will have to wait until https://github.com/openedx/openedx-events/pull/143 is merged and latest version is pushed to pypi. Apologies for this confusion as I was not aware of this restriction.
Do you think we need to revert the changes and create new MR once https://github.com/openedx/openedx-events/pull/143 is merged?

**Comment by sameenfatima78** [comment]
> @sameenfatima78 Ohh, it seems we will have to wait until [openedx/openedx-events#143](https://github.com/openedx/openedx-events/pull/143) is merged and latest version is pushed to pypi. Apologies for this confusion as I was not aware of this restriction. Do you think we need to revert the changes and create new MR once [openedx/openedx-events#143](https://github.com/openedx/openedx-events/pull/143) is merged?

Yeah reverting it makes sense as it could block others from publishing a release.

**Comment by navinkarkera** [comment]
@sameenfatima78 Created #136 MR to revert changes depending on https://github.com/openedx/openedx-events/pull/143. Please approve and merge.

**Pull request merged**